src/12_integrate_HIV.py

```diff
@@ -31,8 +32,21 @@
 cmp1 = cmpraw.filter(F.col('smiles').isNotNull())
 cmp2 = cmp1.withColumn('sid', F.monotonically_increasing_id().cast('string'))
 
+# UDF to convert SMILES to InChI
+def smiles_to_inchi(smiles):
+    mol = Chem.MolFromSmiles(smiles)
+    if mol:
+        return Chem.MolToInchi(mol)
+    else:
+        return None
+
+smiles_to_inchi_udf = udf(smiles_to_inchi, StringType())
+
+# Add InChI column
+cmp_with_inchi = cmp2.withColumn('inchi', smiles_to_inchi_udf(F.col('smiles')))
+
 # Create the substances table in JSON format
-subjson = cmp2.select("sid", F.to_json(F.struct("smiles", "activity", "HIV_active")).alias("data"))
+subjson = cmp_with_inchi.select("sid", F.to_json(F.struct("smiles", "activity", "HIV_active", "inchi")).alias("data"))
```

Comment:
You should use the column name 'value' instead of 'HIV_active'

Reply:
this is meta data so maybe I don't change the column name? Or I can just not include them in the substance table?

Reply:
it looks like my comment is wrong, but the code does still look incorrect. I think you want all of the metadata associated with the substance, not information associated with the property or activity. So it should be something like:

```
subjson = cmp_with_inchi.select("sid", F.to_json(F.struct("smiles", "inchi")).alias("data"))
```